stackslib/src/chainstate/burn/atc.rs

```diff
@@ -0,0 +1,1403 @@
+    pub fn max() -> Self {
+        Self((Uint256::from_u64(u64::MAX) << 64) | Uint256::from_u64(u64::MAX))
+    }
+
+    /// Get integer part
+    pub fn ipart(&self) -> u64 {
+        (self.0 >> 64).low_u64()
+    }
+
+    /// Checked addition
+    pub fn add(&self, other: &AtcRational) -> Option<Self> {
+        // NOTE: this is always safe since u128::MAX + u128::MAX < Uint256::max()
+        let sum = AtcRational(self.0 + other.0);
+        if sum.0 > Self::max().0 {
+            return None;
+        }
+        Some(sum)
+    }
+
+    /// Checked subtraction
+    pub fn sub(&self, other: &AtcRational) -> Option<Self> {
+        if self.0 < other.0 {
+            return None;
+        }
+        Some(AtcRational(self.0 - other.0))
+    }
+
+    /// Checked multiplication
+    pub fn mul(&self, other: &AtcRational) -> Option<Self> {
+        // NOTE: this is always safe since u128::MAX * u128::MAX < Uint256::max()
+        let prod = AtcRational((self.0 * other.0) >> 64);
+        if prod.0 > Self::max().0 {
+            return None;
+        }
+        Some(prod)
+    }
+
+    /// Minimum of self and other
+    pub fn min(&self, other: &AtcRational) -> Self {
+        if self.0 < other.0 {
+            Self(self.0.clone())
+        } else {
+            Self(other.0.clone())
+        }
+    }
+
+    /// Hex representation of the inner bits
+    pub fn to_hex(&self) -> String {
+        self.0.to_hex_be()
+    }
+
+    /// Inner u256, for conversion to something a BurnSamplePoint can use
+    pub fn into_inner(self) -> Uint256 {
+        self.0
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use stacks_common::util::hash::to_hex;
+    use stacks_common::util::uint::Uint256;
+
+    use crate::chainstate::burn::atc::AtcRational;
+    use crate::chainstate::burn::BlockSnapshot;
+
+    impl AtcRational {
+        /// Convert to f64, and panic on conversion failure
+        pub fn to_f64(&self) -> f64 {
+            let ipart = self.ipart() as f64;
+            let fpart = self.0.low_u64() as f64;
+            ipart + (fpart / (u64::MAX as f64))
+        }
+
+        /// Convert from f64 between 0 and 1, panicking on conversion failure.  Scales up the f64 so that its
+        /// fractional parts reside in the lower 64 bits of the AtcRational.
+        pub fn from_f64_unit(value: f64) -> Self {
+            if value < 0.0 || value >= 1.0 {
+                panic!("only usable for values in [0.0, 1.0) range");
+            }
+
+            // NOTE: this only changes the exponent, not the mantissa.
+            // Moreover, u128::from(u64::MAX) + 1 has f64 representation 0x43f0000000000000, so these conversions are safe.
+            let scaled_value = value * ((u128::from(u64::MAX) + 1) as f64);
+
+            // this is safe, because 0.0 <= value < 1.0, so scaled_value <= u64::MAX
+            let value_u64 = scaled_value as u64;
+            Self(Uint256::from_u64(value_u64))
+        }
+    }
+
+    #[test]
+    fn test_atc_rational() {
```

Comment:
I think these tests need more coverage (particularly, the methods like `max`, `ipart` have no coverage here), and should be compared against expected outputs from _other_ representations. For example, asserting `AtcRational::frac(1, 1) == AtcRational::one()` is useful, but it need to also assert `AtcRational::one().ipart() == 1u64` and `AtcRational::one().fpart() == 0u64`. Similarly, the math functions' results should be compared to f64 results (using the test functions you added `to_f64` and `from_f64_unit` to perform those comparisons).

Reply:
Done